File: video/scripts/generate-tts.ts
import { MsEdgeTTS, OUTPUT_FORMAT } from "msedge-tts";
import { writeFileSync, mkdirSync, existsSync } from "fs";
import { join } from "path";
import { segment1 } from "../../lecture/segment-1";
import { segment2 } from "../../lecture/segment-2";
import { segment3 } from "../../lecture/segment-3";
import { segment4 } from "../../lecture/segment-4";
import { FPS } from "../src/constants";
import type { SlideManifest, SubtitleCue, Manifest } from "../src/types";

const VOICE = "ko-KR-SunHiNeural";
const AUDIO_DIR = join(__dirname, "..", "public", "audio");

const allSegments = [...segment1, ...segment2, ...segment3, ...segment4];

interface WordBoundary {
  Type: string;
  Data: {
    Offset: number;
    Duration: number;
    text: { Text: string };
  };
}

function secToFrame(sec: number): number {
  return Math.round(sec * FPS);
}

async function generateSlideAudio(
  slideId: string,
  narration: string,
): Promise<SlideManifest> {
  const tts = new MsEdgeTTS();
  await tts.setMetadata(VOICE, OUTPUT_FORMAT.AUDIO_24KHZ_96KBITRATE_MONO_MP3);

  const audioPath = join(AUDIO_DIR, `${slideId}.mp3`);
  const boundaries: WordBoundary[] = [];

  const readable = tts.toStream(narration);

  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = [];

    readable.on("data", (chunk: Buffer) => {
      chunks.push(chunk);
    });

    readable.on("metadata", (data: WordBoundary) => {
      if (data.Type === "WordBoundary" || data.Type === "SentenceBoundary") {
        boundaries.push(data);
      }
    });

    readable.on("end", () => {
      const buffer = Buffer.concat(chunks);
      writeFileSync(audioPath, buffer);

      // Estimate duration from MP3 bitrate (96kbps = 12000 bytes/sec)
      const durationSec = buffer.length / 12000;
      const durationFrames = secToFrame(durationSec);

      // Convert word boundaries to subtitle cues (sentence-level grouping)
      const cues: SubtitleCue[] = [];
      let sentenceText = "";
      let sentenceStart = -1;
      let sentenceEnd = 0;

      for (const b of boundaries) {
        const offsetSec = b.Data.Offset / 10_000_000;
        const durSec = b.Data.Duration / 10_000_000;
        const startFrame = secToFrame(offsetSec);
        const endFrame = secToFrame(offsetSec + durSec);

        if (b.Type === "SentenceBoundary") {
          // Flush previous sentence
          if (sentenceText && sentenceStart >= 0) {
            cues.push({ text: sentenceText.trim(), startFrame: sentenceStart, endFrame: sentenceEnd });
          }
          sentenceText = b.Data.text.Text;
          sentenceStart = startFrame;
          sentenceEnd = endFrame;
        } else {
          if (sentenceStart < 0) {
            sentenceStart = startFrame;
          }
          sentenceEnd = endFrame;
          if (!sentenceText) {
            sentenceText = b.Data.text.Text;
          }
        }
      }

      // Flush last sentence
      if (sentenceText && sentenceStart >= 0) {
        cues.push({ text: sentenceText.trim(), startFrame: sentenceStart, endFrame: sentenceEnd });
      }

      // If no boundaries captured, create a single cue
      if (cues.length === 0) {
        const sentences = narration.split(/[.!?]\s+/).filter(Boolean);
        const framesPerSentence = Math.floor(durationFrames / Math.max(sentences.length, 1));
        sentences.forEach((s, i) => {
          cues.push({
            text: s.trim(),
            startFrame: i * framesPerSentence,
            endFrame: (i + 1) * framesPerSentence,
          });
        });
      }

      resolve({
        slideId,
        title: slideId,
        audioFile: `audio/${slideId}.mp3`,
        durationSec: Math.round(durationSec * 10) / 10,
        durationFrames,
        cues,
      });
    });

    readable.on("error", reject);
  });
}

async function main() {
  if (!existsSync(AUDIO_DIR)) {
    mkdirSync(AUDIO_DIR, { recursive: true });
  }

  console.log(`Generating TTS for ${allSegments.length} slides...`);

  const slides: SlideManifest[] = [];

  for (const seg of allSegments) {
    console.log(`  ${seg.slideId}: ${seg.narration.length}자...`);
    const result = await generateSlideAudio(seg.slideId, seg.narration);
    result.title = seg.title;
    slides.push(result);
    console.log(`    → ${result.durationSec}s (${result.durationFrames} frames, ${result.cues.length} cues)`);
  }

  const manifest: Manifest = {
    slides,
    totalDurationFrames: slides.reduce((sum, s) => sum + s.durationFrames, 0),
  };

  const manifestPath = join(AUDIO_DIR, "manifest.json");
  writeFileSync(manifestPath, JSON.stringify(manifest, null, 2));

  const totalMin = Math.round(manifest.totalDurationFrames / FPS / 60 * 10) / 10;
  console.log(`\nDone! ${slides.length} slides, ${totalMin} minutes total`);
  console.log(`Manifest: ${manifestPath}`);
}

main().catch(console.error);

File: video/scripts/generate-tts.py
"""Generate TTS audio + subtitle metadata for all lecture segments."""
import asyncio
import json
import os
import sys

import edge_tts

VOICE = "ko-KR-SunHiNeural"
FPS = 30
AUDIO_DIR = os.path.join(os.path.dirname(__file__), "..", "public", "audio")

# Parse lecture TS files to extract segments
def parse_segments():
    """Read all segment-*.ts files and extract narration data."""
    import re
    lecture_dir = os.path.join(os.path.dirname(__file__), "..", "..", "lecture")
    segments = []

    filenames = [f"segment-{i}.ts" for i in range(1, 5)] + ["segment-new.ts"]
    for filename in filenames:
        filepath = os.path.join(lecture_dir, filename)
        with open(filepath, "r") as f:
            content = f.read()

        # Find slideId
        slide_ids = re.findall(r'slideId:\s*"([^"]+)"', content)
        # Find title
        titles = re.findall(r'title:\s*"([^"]+)"', content)
        # Find narration (handles both "..." and `...` strings)
        narrations = re.findall(r'narration:\s*(?:`([\s\S]*?)`|"([\s\S]*?)")\s*,', content)

        for idx, slide_id in enumerate(slide_ids):
            title = titles[idx] if idx < len(titles) else slide_id
            if idx < len(narrations):
                # narrations[idx] is a tuple: (backtick_match, quote_match)
                narration = narrations[idx][0] or narrations[idx][1]
            else:
                narration = ""
            # Clean up
            narration = narration.replace('\\"', '"').replace("\\'", "'").strip()
            if narration:
                segments.append({"slideId": slide_id, "title": title, "narration": narration})

    return segments


def sec_to_frame(sec: float) -> int:
    return round(sec * FPS)


async def generate_slide(slide_id: str, narration: str) -> dict:
    """Generate TTS audio and collect word boundary metadata."""
    audio_path = os.path.join(AUDIO_DIR, f"{slide_id}.mp3")
    communicate = edge_tts.Communicate(narration, VOICE)

    cues = []

    with open(audio_path, "wb") as f:
        async for chunk in communicate.stream():
            if chunk["type"] == "audio":
                f.write(chunk["data"])
            elif chunk["type"] == "SentenceBoundary":
                offset_sec = chunk["offset"] / 10_000_000
                duration_sec = chunk["duration"] / 10_000_000
                text = chunk["text"]
                cues.append({
                    "text": text.strip(),
                    "startFrame": sec_to_frame(offset_sec),
                    "endFrame": sec_to_frame(offset_sec + duration_sec),
                })

    # Get actual audio duration from file size (MP3 96kbps ≈ 12000 bytes/sec)
    file_size = os.path.getsize(audio_path)
    duration_sec = file_size / 12000
    duration_frames = sec_to_frame(duration_sec)

    return {
        "audioFile": f"audio/{slide_id}.mp3",
        "durationSec": round(duration_sec, 1),
        "durationFrames": duration_frames,
        "cues": cues,
    }


async def main():
    os.makedirs(AUDIO_DIR, exist_ok=True)
    segments = parse_segments()

    if not segments:
        print("ERROR: No segments parsed. Check lecture/*.ts files.")
        sys.exit(1)

    print(f"Generating TTS for {len(segments)} slides...")

    slides = []
    for seg in segments:
        sid = seg["slideId"]
        print(f"  {sid}: {len(seg['narration'])}자...", end="", flush=True)
        result = await generate_slide(sid, seg["narration"])
        result["slideId"] = sid
        result["title"] = seg["title"]
        slides.append(result)
        print(f" → {result['durationSec']}s ({result['durationFrames']} frames, {len(result['cues'])} cues)")

    manifest = {
        "slides": slides,
        "totalDurationFrames": sum(s["durationFrames"] for s in slides),
    }

    manifest_path = os.path.join(AUDIO_DIR, "manifest.json")
    with open(manifest_path, "w") as f:
        json.dump(manifest, f, ensure_ascii=False, indent=2)

    total_min = round(manifest["totalDurationFrames"] / FPS / 60, 1)
    print(f"\nDone! {len(slides)} slides, {total_min} minutes total")
    print(f"Manifest: {manifest_path}")


if __name__ == "__main__":
    asyncio.run(main())
